### PR TITLE
Add support for NumPy 2.0

### PR DESCRIPTION
**Pull request by asmeurer**
Fixes https://github.com/data-apis/array-api-compat/issues/77

### CODE REVIEW

**Comment by asmeurer** [comment]
OK, it looks like this is working. This is ready for review. Then I can double check the cupy tests and cut a release.

**Comment by rgommers** [comment]
Two CI failures:
- It looks like ` array-api-tests-numpy-dev / tests (3.8)` needs skipping, since Python 3.8 is too old - that job only picks up numpy 1.24.
- The other failure is a one-off failure for a `__ne__` test due to hypothesis non-reproducibility (could be ignored)

File: numpy-dev-xfails.txt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+array_api_tests/test_has_names.py::test_has_names[array_method-to_device]
+array_api_tests/test_has_names.py::test_has_names[array_attribute-device]
+
+# linalg tests require https://github.com/data-apis/array-api-tests/pull/101
+# cleanups. Also some tests are using .mT
+array_api_tests/test_linalg.py::test_solve
+
+# Array methods and attributes not already on np.ndarray cannot be wrapped
+array_api_tests/test_signatures.py::test_array_method_signature[__array_namespace__]
+array_api_tests/test_signatures.py::test_array_method_signature[to_device]
```
Comment (rgommers) [inline numpy-dev-xfails.txt:17]: About to land: https://github.com/numpy/numpy/pull/25233

File: numpy-dev-xfails.txt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+# Array methods and attributes not already on np.ndarray cannot be wrapped
+array_api_tests/test_has_names.py::test_has_names[array_method-to_device]
+array_api_tests/test_has_names.py::test_has_names[array_attribute-device]
+
+# linalg tests require https://github.com/data-apis/array-api-tests/pull/101
+# cleanups. Also some tests are using .mT
+array_api_tests/test_linalg.py::test_solve
+
+# Array methods and attributes not already on np.ndarray cannot be wrapped
+array_api_tests/test_signatures.py::test_array_method_signature[__array_namespace__]
```
Comment (rgommers) [inline numpy-dev-xfails.txt:16]: This method exists already for a few weeks, and I just merged what's probably the fix to make this test pass: https://github.com/numpy/numpy/pull/25595

**Review by rgommers — APPROVED** [review]
All LGTM, thanks Aaron! 

I added a few comments that may be worth checking and tweaking the comments for for future reference. None are blocking though.

**Comment by asmeurer** [comment]
I guess the nightly builds are lagged. Going to just merge this now to avoid further delay.

**Pull request merged**